In [1]:
import pandas as pd
import numpy as np

In [2]:
path_content = "/Users/Savelyev_Ilya/Downloads/Content.csv"

In [3]:
path_reactions = "/Users/Savelyev_Ilya/Downloads/Reactions.csv"

In [4]:
path_reaction_types = "/Users/Savelyev_Ilya/Downloads/ReactionTypes.csv"

<b>Content Dataset Cleaning</b>

Loading the Content dataset

In [104]:
df = pd.read_csv(path_content, index_col = 0)

In [105]:
df.head()

Content ID                               User ID  \
0  97522e57-d9ab-4bd6-97bf-c24d952602d2  8d3cd87d-8a31-4935-9a4f-b319bfe05f31   
1  9f737e0a-3cdd-4d29-9d24-753f4e3be810  beb1f34e-7870-46d6-9fc7-2e12eb83ce43   
2  230c4e4d-70c3-461d-b42c-ec09396efb3f  a5c65404-5894-4b87-82f2-d787cbee86b4   
3  356fff80-da4d-4785-9f43-bc1261031dc6  9fb4ce88-fac1-406c-8544-1a899cee7aaf   
4  01ab84dd-6364-4236-abbb-3f237db77180  e206e31b-5f85-4964-b6ea-d7ee5324def1   

    Type        Category                                                URL  
0  photo        Studying  https://socialbuzz.cdn.com/content/storage/975...  
1  photo  healthy eating  https://socialbuzz.cdn.com/content/storage/9f7...  
2  photo  healthy eating  https://socialbuzz.cdn.com/content/storage/230...  
3  photo      technology  https://socialbuzz.cdn.com/content/storage/356...  
4  video            food  https://socialbuzz.cdn.com/content/storage/01a...

Dropping User ID and URL columns as they do not seems valuable for the future analysis in our case

In [106]:
df = df.drop(['User ID', 'URL'], axis = 1)

Checking the size of the dataset, looking for null values and making sure all the columns have the appropriate data types

In [107]:
df.shape

(1000, 3)

In [108]:
df.isnull().sum(axis = 0)

Content ID    0
Type          0
Category      0
dtype: int64

In [109]:
df.dtypes

Content ID    object
Type          object
Category      object
dtype: object

Let's make sure that all the values in the columns are unique

In [110]:
df["Type"].unique()

array(['photo', 'video', 'GIF', 'audio'], dtype=object)

In [111]:
df["Category"].unique()

array(['Studying', 'healthy eating', 'technology', 'food', 'cooking',
       'dogs', 'soccer', 'public speaking', 'science', 'tennis', 'travel',
       'fitness', 'education', 'studying', 'veganism', 'Animals',
       'animals', 'culture', '"culture"', 'Fitness', '"studying"',
       'Veganism', '"animals"', 'Travel', '"soccer"', 'Education',
       '"dogs"', 'Technology', 'Soccer', '"tennis"', 'Culture', '"food"',
       'Food', '"technology"', 'Healthy Eating', '"cooking"', 'Science',
       '"public speaking"', '"veganism"', 'Public Speaking', '"science"'],
      dtype=object)

It seems like there are duplicated categories with and without quotations marks. Lets get rid of quotes

In [112]:
df["Category"].replace('"', '', inplace=True, regex=True)

There are also some categories with capital letters. Let's make all the categories written in lower case

In [113]:
df["Category"] = df["Category"].str.lower()

Now let's make sure that all the categories are finally uniqie

In [114]:
df["Category"].unique()

array(['studying', 'healthy eating', 'technology', 'food', 'cooking',
       'dogs', 'soccer', 'public speaking', 'science', 'tennis', 'travel',
       'fitness', 'education', 'veganism', 'animals', 'culture'],
      dtype=object)

Just to make sure there are no duplicates

In [116]:
df.duplicated().sum()

0

Final check of the cleaned dataset and saving it to csv file

In [117]:
df

Content ID   Type         Category
0    97522e57-d9ab-4bd6-97bf-c24d952602d2  photo         studying
1    9f737e0a-3cdd-4d29-9d24-753f4e3be810  photo   healthy eating
2    230c4e4d-70c3-461d-b42c-ec09396efb3f  photo   healthy eating
3    356fff80-da4d-4785-9f43-bc1261031dc6  photo       technology
4    01ab84dd-6364-4236-abbb-3f237db77180  video             food
..                                    ...    ...              ...
995  b4cef9ef-627b-41d7-a051-5961b0204ebb  video  public speaking
996  7a79f4e4-3b7d-44dc-bdef-bc990740252c    GIF       technology
997  435007a5-6261-4d8b-b0a4-55fdc189754b  audio         veganism
998  4e4c9690-c013-4ee7-9e66-943d8cbd27b7    GIF          culture
999  75d6b589-7fae-4a6d-b0d0-752845150e56  audio       technology

[1000 rows x 3 columns]

In [118]:
df.to_csv('Content_cleaned.csv')

<b>Reactions Dataset Cleaning</b>

Loading the Reactions dataset

In [119]:
df1 = pd.read_csv(path_reactions, index_col = 0)
df1.head()

Content ID                               User ID  \
0  97522e57-d9ab-4bd6-97bf-c24d952602d2                                   NaN   
1  97522e57-d9ab-4bd6-97bf-c24d952602d2  5d454588-283d-459d-915d-c48a2cb4c27f   
2  97522e57-d9ab-4bd6-97bf-c24d952602d2  92b87fa5-f271-43e0-af66-84fac21052e6   
3  97522e57-d9ab-4bd6-97bf-c24d952602d2  163daa38-8b77-48c9-9af6-37a6c1447ac2   
4  97522e57-d9ab-4bd6-97bf-c24d952602d2  34e8add9-0206-47fd-a501-037b994650a2   

      Type             Datetime  
0      NaN  2021-04-22 15:17:15  
1  disgust  2020-11-07 09:43:50  
2  dislike  2021-06-17 12:22:51  
3   scared  2021-04-18 05:13:58  
4  disgust  2021-01-06 19:13:01

Dropping User ID column as we did with the previous dataset. The Datetime column may be valuable in the future analysis, thus we will keep it here for now

In [120]:
df1 = df1.drop(['User ID'], axis = 1)
df1.head()

Content ID     Type             Datetime
0  97522e57-d9ab-4bd6-97bf-c24d952602d2      NaN  2021-04-22 15:17:15
1  97522e57-d9ab-4bd6-97bf-c24d952602d2  disgust  2020-11-07 09:43:50
2  97522e57-d9ab-4bd6-97bf-c24d952602d2  dislike  2021-06-17 12:22:51
3  97522e57-d9ab-4bd6-97bf-c24d952602d2   scared  2021-04-18 05:13:58
4  97522e57-d9ab-4bd6-97bf-c24d952602d2  disgust  2021-01-06 19:13:01

Checking the size of the dataset, looking for null values and making sure all the columns have the appropriate data types

In [121]:
df1.shape

(25553, 3)

In [122]:
df1.dtypes

Content ID    object
Type          object
Datetime      object
dtype: object

It seems like Datetime column has the inappropriate data format. Let's fix it

In [124]:
df1["Datetime"] = pd.to_datetime(df1["Datetime"])

Let's make sure the changes were successful

In [128]:
df1.dtypes

Content ID            object
Type                  object
Datetime      datetime64[ns]
dtype: object

In [130]:
df1.isnull().sum(axis = 0)

Content ID      0
Type          980
Datetime        0
dtype: int64

There are 980 missing values in the Type column. Since the task prescripts us to drop all the rows with null values I will just do "dropna" and make sure they were dropped successfully

In [131]:
df1 = df1.dropna(axis = 0)
df1.shape

(24573, 3)

Now let's make sure that all the values in the Type column are unique

In [132]:
df1['Type'].unique()

array(['disgust', 'dislike', 'scared', 'interested', 'peeking', 'cherish',
       'hate', 'love', 'indifferent', 'super love', 'intrigued',
       'worried', 'like', 'heart', 'want', 'adore'], dtype=object)

Seems like everything is fine! Lastly I will make sure, that there are now duplicated rows in the dataset

In [134]:
df1.duplicated().sum()

0

Great! Final check of the cleaned dataset and saving it to csv file

In [135]:
df1

Content ID        Type            Datetime
1      97522e57-d9ab-4bd6-97bf-c24d952602d2     disgust 2020-11-07 09:43:50
2      97522e57-d9ab-4bd6-97bf-c24d952602d2     dislike 2021-06-17 12:22:51
3      97522e57-d9ab-4bd6-97bf-c24d952602d2      scared 2021-04-18 05:13:58
4      97522e57-d9ab-4bd6-97bf-c24d952602d2     disgust 2021-01-06 19:13:01
5      97522e57-d9ab-4bd6-97bf-c24d952602d2  interested 2020-08-23 12:25:58
...                                     ...         ...                 ...
25548  75d6b589-7fae-4a6d-b0d0-752845150e56     dislike 2020-06-27 09:46:48
25549  75d6b589-7fae-4a6d-b0d0-752845150e56   intrigued 2021-02-16 17:17:02
25550  75d6b589-7fae-4a6d-b0d0-752845150e56  interested 2020-09-12 03:54:58
25551  75d6b589-7fae-4a6d-b0d0-752845150e56     worried 2020-11-04 20:08:31
25552  75d6b589-7fae-4a6d-b0d0-752845150e56     cherish 2021-01-04 04:55:11

[24573 rows x 3 columns]

In [136]:
df1.to_csv('Reactions_cleaned.csv')

<b>ReactionTypes Dataset Cleaning</b>

Loading the ReactionTypes dataset

In [137]:
df2 = pd.read_csv(path_reaction_types, index_col = 0)
df2.head()

Type Sentiment  Score
0       heart  positive     60
1        want  positive     70
2     disgust  negative      0
3        hate  negative      5
4  interested  positive     30

I don't see any columns to drop from this dataset, as the Sentiment column may also be useful for simplification. So, we leave the columns as it is and check the size of the dataset, looking for null values and making sure all the columns have the appropriate data types

In [138]:
df2.shape

(16, 3)

In [139]:
df2.dtypes

Type         object
Sentiment    object
Score         int64
dtype: object

In [140]:
df2.isnull().sum(axis = 0)

Type         0
Sentiment    0
Score        0
dtype: int64

For now everything is fine! Now I will check all the unique values in Type and Sentiment columns

In [141]:
df2['Type'].unique()

array(['heart', 'want', 'disgust', 'hate', 'interested', 'indifferent',
       'love', 'super love', 'cherish', 'adore', 'like', 'dislike',
       'intrigued', 'peeking', 'scared', 'worried'], dtype=object)

In [142]:
df2['Sentiment'].unique()

array(['positive', 'negative', 'neutral'], dtype=object)

Great. Finally let's look for any duplicated rows

In [143]:
df2.duplicated().sum()

0

Cool! Final check of the cleaned dataset and saving it to csv file

In [144]:
df2

Type Sentiment  Score
0         heart  positive     60
1          want  positive     70
2       disgust  negative      0
3          hate  negative      5
4    interested  positive     30
5   indifferent   neutral     20
6          love  positive     65
7    super love  positive     75
8       cherish  positive     70
9         adore  positive     72
10         like  positive     50
11      dislike  negative     10
12    intrigued  positive     45
13      peeking   neutral     35
14       scared  negative     15
15      worried  negative     12

In [145]:
df2.to_csv('ReactionTypes_cleaned.csv')

Now we need to join all three cleaned dataframes together. First of all I will join Content dataframe to Reactions with the left type of join based on Content ID match

In [180]:
df_join = pd.merge(df1, df, how = 'left', on = 'Content ID')

In [181]:
df_join

Content ID      Type_x            Datetime  \
0      97522e57-d9ab-4bd6-97bf-c24d952602d2     disgust 2020-11-07 09:43:50   
1      97522e57-d9ab-4bd6-97bf-c24d952602d2     dislike 2021-06-17 12:22:51   
2      97522e57-d9ab-4bd6-97bf-c24d952602d2      scared 2021-04-18 05:13:58   
3      97522e57-d9ab-4bd6-97bf-c24d952602d2     disgust 2021-01-06 19:13:01   
4      97522e57-d9ab-4bd6-97bf-c24d952602d2  interested 2020-08-23 12:25:58   
...                                     ...         ...                 ...   
24568  75d6b589-7fae-4a6d-b0d0-752845150e56     dislike 2020-06-27 09:46:48   
24569  75d6b589-7fae-4a6d-b0d0-752845150e56   intrigued 2021-02-16 17:17:02   
24570  75d6b589-7fae-4a6d-b0d0-752845150e56  interested 2020-09-12 03:54:58   
24571  75d6b589-7fae-4a6d-b0d0-752845150e56     worried 2020-11-04 20:08:31   
24572  75d6b589-7fae-4a6d-b0d0-752845150e56     cherish 2021-01-04 04:55:11   

      Type_y    Category  
0      photo    studying  
1      photo    studying  
2      photo    studying  
3      photo    studying  
4      photo    studying  
...      ...         ...  
24568  audio  technology  
24569  audio  technology  
24570  audio  technology  
24571  audio  technology  
24572  audio  technology  

[24573 rows x 5 columns]

In both datasets there was a Type column, thus pandas changed their names to distinguish them. Since I do not really like how they look now, I apply new names to them by myself

In [182]:
df_join.rename(columns = {"Type_x" : "Type", "Type_y": "Content Type"}, inplace = True)

Now I am going to join the ReactionTypes dataset to the new dataset my mathching Type columns

In [184]:
df_joined = pd.merge(df_join, df2, on = 'Type', how = 'left')

In [186]:
df_joined

Content ID        Type            Datetime  \
0      97522e57-d9ab-4bd6-97bf-c24d952602d2     disgust 2020-11-07 09:43:50   
1      97522e57-d9ab-4bd6-97bf-c24d952602d2     dislike 2021-06-17 12:22:51   
2      97522e57-d9ab-4bd6-97bf-c24d952602d2      scared 2021-04-18 05:13:58   
3      97522e57-d9ab-4bd6-97bf-c24d952602d2     disgust 2021-01-06 19:13:01   
4      97522e57-d9ab-4bd6-97bf-c24d952602d2  interested 2020-08-23 12:25:58   
...                                     ...         ...                 ...   
24568  75d6b589-7fae-4a6d-b0d0-752845150e56     dislike 2020-06-27 09:46:48   
24569  75d6b589-7fae-4a6d-b0d0-752845150e56   intrigued 2021-02-16 17:17:02   
24570  75d6b589-7fae-4a6d-b0d0-752845150e56  interested 2020-09-12 03:54:58   
24571  75d6b589-7fae-4a6d-b0d0-752845150e56     worried 2020-11-04 20:08:31   
24572  75d6b589-7fae-4a6d-b0d0-752845150e56     cherish 2021-01-04 04:55:11   

      Content Type    Category Sentiment  Score  
0            photo    studying  negative      0  
1            photo    studying  negative     10  
2            photo    studying  negative     15  
3            photo    studying  negative      0  
4            photo    studying  positive     30  
...            ...         ...       ...    ...  
24568        audio  technology  negative     10  
24569        audio  technology  positive     45  
24570        audio  technology  positive     30  
24571        audio  technology  negative     12  
24572        audio  technology  positive     70  

[24573 rows x 7 columns]

Great job! Will save it to the new csv file for now

In [187]:
df_joined.to_csv('FinalDataset.csv')

The task was to group the dataset by categories of the content and choose the top 5 categories by total sum of scores. Let's do it!

In [190]:
df_cat = df_joined.groupby("Category").sum('Score')
df_cat

Score
Category              
animals          74965
cooking          64756
culture          66579
dogs             52511
education        57436
fitness          55323
food             66676
healthy eating   69339
public speaking  49264
science          71168
soccer           57783
studying         54269
technology       68738
tennis           50339
travel           64880
veganism         49619

In [192]:
df_cat_sorted = df_cat.sort_values(by = 'Score', ascending = False)
df_cat_sorted

Score
Category              
animals          74965
science          71168
healthy eating   69339
technology       68738
food             66676
culture          66579
travel           64880
cooking          64756
soccer           57783
education        57436
fitness          55323
studying         54269
dogs             52511
tennis           50339
veganism         49619
public speaking  49264

In [193]:
df_cat_top5 = df_cat_sorted.head()
df_cat_top5

Score
Category             
animals         74965
science         71168
healthy eating  69339
technology      68738
food            66676

Great! I finally save this new dataframe to a csv file format

In [194]:
df_cat_top5.to_csv('Top5.csv')